# TMA01, question 1 (35 marks)

**Name**: Simon Malpas

**PI**: B3944330

In this question, you will investigate a dataset of opera performances. You are interested in the question:

- Where do operatic performances take place in the United Kingdom?

To address this question, you want to plot a map of the United Kingdom, with markers representing the locations of opera performances, with the size of the marker indicating the number of performances which took place at that location.

For this task, you will only consider the 2017-2018 opera season.

<img src="images/covent_garden.jpg" alt="By Thomas Rowlandson (1756–1827) and Augustus Charles Pugin (1762–1832) (after) John Bluck (fl. 1791–1819), Joseph Constantine Stadler (fl. 1780–1812), Thomas Sutherland (1785–1838), J. Hill, and Harraden (engravers)" style="width: 400px;"/>

<p style="text-align: center;">The Covent Garden Theatre (<a href="https://commons.wikimedia.org/w/index.php?curid=15799255)">Public Domain from wikimedia.org</a>)
    



The tasks in this notebook can be addressed using the techniques discussed in parts 2-6 of the module materials and the associated notebooks.

The question has three parts, looking at different parts of the data analysis pipeline.

Record all your activity and observations in this notebook. Insert additional notebook cells as required. Remember to run each cell in sequence and to rerun cells if you make any changes in earlier cells. 

Before you submit your notebook make sure you run all cells in order and check that you get the results you expect. (It is not unknown to receive notebooks which don't work when the cells are run in order. The most reliable way of checking your results is usually to use the menu option *Kernel $\rightarrow$ Restart & Run All*.)

Note that in this question you are required to use Python and the pandas library - this is to give you experience with using pandas and DataFrames to manipulate data.

**Note:** You should be prepared to use [the pandas documentation](https://pandas.pydata.org/docs/) to find the most appropriate functions or methods for your analysis.

In [1]:
# This cell imports the standard pandas library.

import pandas as pd

# You will also need to plot a map, so import folium:

import folium

#### Contents

[Data provenance and importing the data](#provenance)

[Cleaning, reshaping and combining the data sets](#combining)

[Visualising the data](#visualising)

## <a id='provenance'></a>Data provenance, importing and shaping the data

In this notebook, you will use two datasets. You can find these in the `data` directory. Although we have provided both for you here, even when someone passes you a dataset, you need to be able to confirm your usage rights for that data.

#### 1. Licensing for the Opera dataset

The dataset for the 2017-2018 opera season is part of the larger dataset:

    Cuntz, Alexander, 2020, "Replication Data for: Grand rights and opera reuse today",    
    https://doi.org/10.7910/DVN/8LUFN8, Harvard Dataverse, V1

It is stored as a csv file called `stats1718.csv` in the `data` directory. This dataset was obtained from the [Harvard dataverse portal](https://dataverse.harvard.edu/) on 13th March, 2021 from:

https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/8LUFN8


Find the terms under which the dataset can be used, and state, with a quote and reference:
    
1. the specific statements which allow the OU to distribute the data to you, and
1. the obligations placed on the OU when distributing the data.

*(2 marks)*


**Write your answer in this markdown cell**
1.
On the page where the dataset can be downloaded, there is a box called 'Terms'
Within this the waiver states "CC0 - "Public Domain Dedication" "

Following through to the community norms, and then creative commons (the licensing organisation), the license type is clarified as:

CC0
“No Rights Reserved”


CC0 enables scientists, educators, artists and other creators and owners of copyright- or database-protected content to waive those interests in their works and thereby place them as completely as possible in the public domain, so that others may freely build upon, enhance and reuse the works for any purposes without restriction under copyright or database law.
https://creativecommons.org/share-your-work/public-domain/cc0

2.
Following this information regarding the license, there are no legal obligations placed on the OU when distributing this data, however the Dataverse Community Norms, ask that data is attributed (citation), and not used to identify individuals. 


#### 2. Licensing for the City dataset

The information about towns and cities in the world which have a population of more than 1000 people is stored as a text file in the `data` directory called `cities1000.txt`. This dataset was obtained from the [GeoNames geographical database](https://www.geonames.org/) page:

https://download.geonames.org/export/dump/

on 13th March, 2021.

The `cities1000.txt` file is governed by the [Creative Commons Attribution 4.0](https://creativecommons.org/licenses/by/4.0/) licence.


Find the details of this licence, and state:
1. the clauses of the licence which allow the OU to distribute the data to you, and
2. the obligations that the licence places upon the OU when distributing the data.

How do you know that the version in the `data` directory is the same as the one on the GeoNames site (on the day of download)?

*(3 marks)*



1. The license is very simple, it states:

You are free to:

Share — copy and redistribute the material in any medium or format.
 
Adapt — remix, transform, and build upon the material
for any purpose, even commercially. 

This means that the OU (indeed anyone) can do what ever they like with the data, including use it in a way that makes money

2.
Again, it very simply states:

Attribution — You must give appropriate credit, provide a link to the license, and indicate if changes were made. You may do so in any reasonable manner, but not in any way that suggests the licensor endorses you or your use. 

This means the OU needs to make it clear where they obtained the data, who created it, and if they have changed it in any way.

3.
Whilst the last modified timestamp on the website, and comparing file sizes would give an indication of the provinence of the data, it is not conclusive. The method of comparing checksums allows knowledge that the downloaded file is the same data as that held by the website. This would involve the data provider/host to calculate the checksum, publish it along with the method so that the checksum can be duplicated following download and compared. This is more conclusive, but still involves trust of the website providing the data and the downloader's network environment.

#### 3. Importing the opera dataset

Each row of the opera dataset represents a run of performances of a particular opera in a given location. The details of the columns are given in the *notes* on the [dataset's host page](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/8LUFN8).

Import the file `stats1718.csv`, and create a DataFrame named `uk_opera_df` which contains:
- a column representing the town or city where the opera was performed, and
- a column containing the number of performances of the opera in the run.

The DataFrame should only contain those operas which were performed in the United Kingdom. In the file `stats1718.csv`, performances in the United Kingdom are represented by the two letter ISO code `uk`.

You should ensure that the columns in your DataFrame `uk_operas_df` have meaningful names.

Display a preview of the first five rows of the DataFrame.

*(4 marks)*

In [2]:
#have a look at the start of the csv to see the foramtting of the file
! head data/stats1718.csv
# I also confirmed this by loading the csv into openrefine, seeing that there are 17 columns recognised,
# meaning column names will have to be provided for the empty columns.

In [3]:
#I notice from this that the separators are a pipe symbol, this will need to be set in the read_csv function of pandas
# also there is no column header included in the file, so this will have to be added manually from the data on the
#website that hosts the dataset.

columnNames = ['season', 'iso','three', 'town', 'five', 'composer', 'db', 'dd',\
               'nat', 'mf', 'work', 'work_nat', 'type', 'start_date', 'performances', 'new/concert', 'seventeen']
# read in csv, using defined column names (including blanks), header is none as first line of file is data
all_operas_df = pd.read_csv('data/stats1718.csv', names=columnNames, sep='|', header=None)

all_operas_df.head()

season iso  three     town  five    composer        db        dd nat mf  \
0    1718  al    NaN   Tirana   NaN       Bizet  18381025  18750603  fr  m   
1    1718  al    NaN   Tirana   NaN       Bizet  18381025  18750603  fr  m   
2    1718  al    NaN   Tirana   NaN      Brahms  18330507  18970403  de  m   
3    1718  al    NaN   Tirana   NaN   Strauss,J      1825      1899  at  m   
4    1718  am    NaN  Yerevan   NaN  Arutiunian  19200928  20120328  am  m   

                     work work_nat type  start_date  performances new/concert  \
0                  Carmen       fr    ~    20180130             5         NaN   
1  Les Pecheurs de perles       fr    ~    20170922             3           !   
2   Ein deutsches Requiem       de   ~O    20170930             1           c   
3          Die Fledermaus       at   ~L    20180329             5         NaN   
4              Sayat-Nova       am    ~    20180313             1         NaN   

   seventeen  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN

Select the operas performed in the uk by filtering by column.

Then select columns for town and number of performances.

Finally reindex the dataframe, and delete (drop) the original index numbering.

In [4]:
#this could be done with a single line of code, but easier to follow if split into


uk_opera_df = all_operas_df[all_operas_df["iso"] == 'uk'][['town', 'performances']].reset_index(drop=True)

#display first 5 rows
uk_opera_df.head(5)

town  performances
0     Aberdeen             2
1     Aberdeen             2
2      Bampton             2
3       Bangor             1
4  Basingstoke             1

#### 4. Importing the cities dataset

Each row of the cities dataset represents a town or city with a population of more than 1000 people. The details of the columns are given on the [dataset's download page](http://download.geonames.org/export/dump/).

Import the file `cities1000.txt`, and create a DataFrame named `uk_cities_df` which contains:
- a column containing the name of the town or city,
- a column containing the city's latitude, and
- a column containing the city's longitude.

You should ensure that the columns in your DataFrame `uk_cities_df` have meaningful names.

The DataFrame should only contain those towns and cities in in the United Kingdom. In the file `cities1000.txt`, towns and cities in the United Kingdom are represented by the two letter ISO code `GB`.

Display a preview of the first five rows of the DataFrame. 

*(4 marks)*

Dataset download page states that the file is tab delimited, and the encoding is UTF8. I will use this along with the stated list of column names given, which I will place in a list of strings to be used as column names for the read_csv function (\t is the char for tab)

In [5]:
# create a list of the column names detailed in the documentation
citiesColumns = ['geonameid', 'name', 'asciiname', 'alternatenames', 'latitude',\
                'longitude', 'featureClass', 'featureCode', 'countryCode', 'cc2',\
                'admin1', 'admin2', 'admin3', 'admin4', 'population', 'elevation',\
                'dem', 'timezone', 'modificationDate']

#import the data, explicitly stating the encoding from the documentation, the tab separator (\t), the column names
# and that there is no header line included in the csv file.
cities_df = pd.read_csv('data/cities1000.txt', encoding='UTF8', sep='\t', names=citiesColumns, header=0)

# display the first rows to check a correct import
cities_df.head(2)

/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3441: DtypeWarning: Columns (13) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


geonameid                 name            asciiname  \
0    3039163  Sant Julià de Lòria  Sant Julia de Loria   
1    3039604       Pas de la Casa       Pas de la Casa   

                                      alternatenames  latitude  longitude  \
0  San Julia,San Julià,Sant Julia de Loria,Sant J...  42.46372    1.49129   
1                      Pas de la Kasa,Пас де ла Каса  42.54277    1.73361   

  featureClass featureCode countryCode  cc2 admin1 admin2 admin3 admin4  \
0            P        PPLA          AD  NaN     06    NaN    NaN    NaN   
1            P         PPL          AD  NaN     03    NaN    NaN    NaN   

   population  elevation   dem        timezone modificationDate  
0        8022        NaN   921  Europe/Andorra       2013-11-23  
1        2363     2050.0  2106  Europe/Andorra       2008-06-09

When attempting to import the data, there is a DtypeWarning. Further research led me to understand that
this is due to the way that python processes the file (in chunks) and that the automatic type recognition
becomes confused when it comes across chunks that are all integers whilst in other chunks they are strings,
or even NaN. By examining the Dtypes of the dataframe following import, and comparing to the types stated 
in the data source, it can be seen that the recognition is valid, except in the case of elevation, which
is recognised as a float, but is stated as an integer.

In [6]:
# data types could have been passed as a dict to the read_csv function by adding dtypes=citiesDtypes
# unnecesary as despite warning, dtypes are as expected.
# citiesDtypes = {'geonameid':'str', \
#                 'name':'str', \
#                 'asciiname':'str', \
#                 'alternatenames':'str', \
#                 'latitude': 'float64',\
#                 'longitude': 'float64', \
#                 'featureClass':'str', \
#                 'featureCode':'str', \
#                 'countryCode':'str', \
#                 'cc2':'str',\
#                 'admin1':'str', \
#                 'admin2':'str', \
#                 'admin3':'str', \
#                 'admin4':'str', \
#                 'population': 'int64', \
#                 'elevation': 'float64',\
#                 'dem': 'int64', \
#                 'timezone':'str', \
#                 'modificationDate':'str'}

cities_df.dtypes

geonameid             int64
name                 object
asciiname            object
alternatenames       object
latitude            float64
longitude           float64
featureClass         object
featureCode          object
countryCode          object
cc2                  object
admin1               object
admin2               object
admin3               object
admin4               object
population            int64
elevation           float64
dem                   int64
timezone             object
modificationDate     object
dtype: object

In [7]:
#filter out just uk cities.
uk_cities_df = cities_df[cities_df['countryCode']== 'GB']

# pull out the desired columns, and overwrite the variable.
uk_cities_df = uk_cities_df[['name', 'latitude', 'longitude']]

# reset the index and drop the original index.
uk_cities_df.reset_index(drop=True, inplace=True)

# show the first 5 entries in the new data frame, noting that the cities are in reverse alphabetical order.
uk_cities_df.head(5)

name  latitude  longitude
0  Ystrad Mynach  51.64276   -3.23620
1  Ystradgynlais  51.76667   -3.76667
2     Ystalyfera  51.76716   -3.78082
3         Yoxall  52.76659   -1.79068
4     Youlgreave  53.17399   -1.69044

## <a id='combining'></a>Cleaning the data sets

You want to combine the information on operatic performances with the geographical information on where the performances took place.

Each row of the the `uk_operas_df` DataFrame contains the name of the town or city where a run of performances took place. Each row of the `uk_cities_df` DataFrame contains the name of the town or city at the location of the given coordinates.

You notice that there are several discrepancies between the values in the two columns representing the towns or cities. These could lead to errors in such an analysis.

#### <a id='identifying_discrepancies'></a>5. Identifying discrepancies between the datasets

When combining datasets, it is important to be systematic about investigating the differences in the columns that you want to use which contain equivalent values. 

Write a statement to generate a set of the town and city names in your DataFrame `uk_opera_df` which do not appear in the DataFrame `uk_cities_df`. 

*(1 mark)*

In [8]:
# Perform a full outer join on the data on the town/name column.

unmatched_towns = uk_opera_df.merge(uk_cities_df, how='outer', left_on='town', right_on='name')

# Extract the rows where there is a key from uk_opera_df, but no matching key from uk_cities_df and the join function
# left the value of the cell in the 'name' column empty i.e. null. This can be checked for using pd.isnull()
unmatched_towns = unmatched_towns[pd.isnull(unmatched_towns['name'])]

# display an array of the unique names. Iterate over the array for neater output.
for i in unmatched_towns['town'].unique():
    print(i, end=', ')

Birmingham UK, Glyndebourne, Hull, Iford, Longborough, Northington, Salford Quays, Snape, Westonbirt, 

I was unsure what was meant by 'statement' (in C type languages I understand this to be a single line of code
terminated by a semicolon, rather than multiple lines of code). Due to this I have also included a method 
performing the trasformation using a single line of code.
By using an indicator column (by default labeled '_merge') to show which rows are present in just one, 
or both of the dataframes. We can then use .loc to filter the rows indicated as 'left_only'
i.e. no matching key present in uk_cities_df

In [9]:

towns_not_listed = uk_opera_df.merge(uk_cities_df, how='outer', left_on='town', right_on='name', indicator=True).loc[lambda x:x['_merge'] == 'left_only']['town'].unique()

# show the resulting array
towns_not_listed

array(['Birmingham UK', 'Glyndebourne', 'Hull', 'Iford', 'Longborough',
       'Northington', 'Salford Quays', 'Snape', 'Westonbirt'],
      dtype=object)

#### 6. Correcting the discrepancies between the datasets

In order to combine the DataFrames `uk_opera_df` and `uk_cities_df` into a single DataFrame, you need to accommodate the variations in the town and city names in these two DataFrames. When investigating the UK opera season, you make the following observations:


1. [Glyndebourne](https://www.glyndebourne.com/) is an opera house near the town of Lewes, so the occurrences of `Glyndebourne` should be altered to `Lewes`.

2. [Iford Arts](https://ifordarts.org.uk/) is a small opera company based in based in Bradford on Avon near the town of Bath, so the occurrences of `Iford` should be altered to `Bradford-on-Avon`.

3. The [Longborough Opera Festival](https://lfo.org.uk/) takes place near the town of Moreton-in-Marsh, so the occurrences of `Longborough` should be altered to `Moreton in Marsh`.

4. [The Grange at Northington](https://thegrangefestival.co.uk/) hosts an annual opera festival in Northington near Winchester, so the occurrences of `Northington` should be altered to `Winchester`.

5. [Snape Maltings](https://snapemaltings.co.uk/) is a member of the European Network of Opera Academies, based in Snape near Aldeburgh, so the occurrences of `Snape` should be altered to `Aldeburgh`.

6. [Bampton Classical Opera](https://www.bamptonopera.org/) takes place at locations in Bampton and Westonbirt near Tetbury, so the occurrences of `Westonbirt` should be altered to `Tetbury`.

Use this information, **and the results of [part 5](#identifying_discrepancies)**, to ensure that the information about the town or city names in the two DataFrames is consistent. Remember to justify any further decisions you make in aligning the values in the DataFrames.

*(8 marks)*

First lets look at cities not mentioned, 'Birmingham UK', 'Salford Quays', and 'Hull'. We can use the series method .contains to search the df, returning only the rows that contain our string. By using a regex we can investigate all three of these and pull up any rows containing these names.

In [10]:
uk_cities_df[uk_cities_df['name'].str.contains('Birmingham|Salford|Hull', regex=True) == True]

name  latitude  longitude
849             Salfords  51.20430   -0.16947
850              Salford  53.48771   -2.29042
1899  Kingston upon Hull  53.74460   -0.33525
3477          Birmingham  52.48142   -1.89983

A quick google confirms that 'Hull' is a common shortening of the town 'Kingston upon Hull'. Birmingham just needs the ' UK' removing (presumably this was added to avoid confusion with the many Birminghams in the US. The coords of Salford confirm that it is the correct entry ('Salfords' is a town near gatwick).

We now have enough information to put all our substitutions into a dictionary with the original data as the key, and the replacement as the value.

In [11]:
replacements = {'Glyndebourne': 'Lewes',
                'Iford' : 'Bradford-on-Avon',
                'Longborough' : 'Moreton in Marsh',
                'Northington' : 'Winchester',
                'Snape' : 'Aldeburgh',
                'Westonbirt' : 'Tetbury',
                'Birmingham UK' : 'Birmingham',
                'Salford Quays' : 'Salford',
                'Hull' : 'Kingston upon Hull'}

Then use the replace command on the data frame defining the column to be modified.

In [12]:
cleaned_towns_opera_df = uk_opera_df.replace(replacements)
cleaned_towns_opera_df

town  performances
0       Aberdeen             2
1       Aberdeen             2
2        Bampton             2
3         Bangor             1
4    Basingstoke             1
..           ...           ...
272       Woking             2
273       Woking             2
274         York             1
275         York             1
276         York             1

[277 rows x 2 columns]

Now we can perform our join on our uk_cities_df as before and check that we have been successful. As the array is empty all the towns in uk_operas_df have been matched with city data.

In [13]:
mismatched_towns = cleaned_towns_opera_df.merge(uk_cities_df, how='outer', left_on='town', right_on='name', indicator=True).loc[lambda x:x['_merge'] == 'left_only']['town'].unique()

mismatched_towns

array([], dtype=object)

Finally lets sort our towns alphabetically, reset out index, then overwrite (reassign) our uk_opera_df variable with our cleaned data frame

In [14]:
uk_opera_df = cleaned_towns_opera_df.sort_values(by='town')
uk_opera_df.reset_index(drop=True, inplace=True)
uk_opera_df

town  performances
0     Aberdeen             2
1     Aberdeen             2
2    Aldeburgh             2
3    Aldeburgh             1
4    Aldeburgh             1
..         ...           ...
272     Woking             2
273     Woking             2
274       York             1
275       York             1
276       York             1

[277 rows x 2 columns]

#### 7. Removing duplicates from the dataset

You discover that two of the opera locations, `Buxton` and `Bangor` are ambiguous; the United Kingdom has more than one town named Buxton, and more than one town named Bangor.

Either update or reconstruct your DataFrame `uk_cities_df`, so that it contains only one entry for `Buxton` and one entry for `Bangor`:

- The entry for `Bangor` should refer to the Welsh town, which has a population of 17,988, rather than the Northern Irish town, which has a population of 60,385. 

- The entry for `Buxton` should refer to the town with a population of 21,300, rather than the village with a population of 1,316.

*(4 marks)*

First lets rebuild out uk_cities data frame, but retaining the population

In [15]:
#filter out just uk cities.
uk_cities_pop_df = cities_df[cities_df['countryCode']== 'GB']

# pull out the desired columns, and overwrite the variable.
uk_cities_pop_df = uk_cities_pop_df[['name', 'latitude', 'longitude', 'population']]

# reset the index and drop the original index.
uk_cities_pop_df.reset_index(drop=True, inplace=True)

Now we can pull out just the rows with 'Bangor' and 'Buxton' in the name column. The regex operator '|' (pipe) symbol means or in this context. The following command uses the string method contains() on each entry in the series created from the 'name' column. This replaces the entry in each cell with True or False depending on the condition defined by the regex. We then only display the rows where the regex evaluates to True. We use the contains method as it allows us to use a regex where a straight comparison would not allow a partial match.

In [16]:
a =  [6,4,9,2,4,3,4,1,3]
a.sort()
print(a)

[1, 2, 3, 3, 4, 4, 4, 6, 9]


In [17]:
uk_cities_pop_df[uk_cities_pop_df['name'].str.contains('Bangor|Buxton', regex=True) == True]

name  latitude  longitude  population
3194  Buxton  53.25741   -1.90982       21300
3195  Buxton  52.75255    1.29982        1316
3616  Bangor  54.66079   -5.66802       60385
3617  Bangor  53.22752   -4.12936       17988

Buxton we desire has index 3195, and the Bangor we want is at index 3617. Lets drop the rows at these indexes, we can modify the df by using the inplace optional argument. Then search again to confirm we were successful.

In [18]:
#Now drop the row with population 60385
uk_cities_pop_df.drop(index=[3195, 3616], inplace=True)
uk_cities_pop_df[uk_cities_pop_df['name'].str.contains('Bangor|Buxton', regex=True) == True]

name  latitude  longitude  population
3194  Buxton  53.25741   -1.90982       21300
3617  Bangor  53.22752   -4.12936       17988

Finally lets drop the population column and save the newly cleaned data frame as our uk_cities_cleaned_df, and check it worked.


In [19]:
uk_cities_cleaned_df = uk_cities_pop_df.drop('population', axis=1)
uk_cities_cleaned_df[uk_cities_cleaned_df['name'].str.contains('Bangor|Buxton|Hull', regex=True) == True]

name  latitude  longitude
1899  Kingston upon Hull  53.74460   -0.33525
3194              Buxton  53.25741   -1.90982
3617              Bangor  53.22752   -4.12936

## <a id='visualising'></a>Visualising the data

You have been asked to answer the question:

- Where do operatic performances take place in the United Kingdom?

Your task is to plot a map of the UK, containing markers where the size of each marker represents the number of opera performances which took place at that location in the 2017/2018 opera season.


#### 8. Create a DataFrame containing plotting data

To create your map, you should create a new DataFrame called `plot_df` that combines the data from your DataFrames `uk_operas_df` and `uk_cities_df`. Your DataFrame should have a row for each town or city, containing the name of the town or city, the latitude and longitude of the location and the total number of opera performances in the season.

For example, if the `uk_opera_df` DataFrame contained the data:

| City | Performances |
|----|----|
|Oxford | 2 |
|Oxford |	3 |
|Oxford	| 2 |
|Lincoln	| 4 |
|Lincoln	| 4 |

and your `uk_cities_df` DataFrame contained the data:

| City | longitude | latitude |
|----|----|----|
|Oxford | 51.75222 | -1.25596 |
|Lincoln	| 53.22683 | -0.53792 |


then your final DataFrame `plot_df` would contain the data:

| City | longitude | latitude | Performances|
|----|----|----|----|
|Oxford | 51.75222 | -1.25596 | 7 |
|Lincoln	| 53.22683 | -0.53792 | 8|

Note that your own DataFrames might look different from these; the important thing is that the data is correctly structured.


When you have constructed your DataFrame, display the first 5 rows.

*(4 marks)*

First we will create a new data frame that aggregates all the performances for each city to keep our original data clean. We do this by grouping the data by city name, then calling sum on the 'performances' column. We then reset the index to make 'town' a column again.

In [20]:
total_perform_df = uk_opera_df.groupby('town').aggregate({'performances': 'sum'})

# make 'town' a column again
total_perform_df.reset_index(inplace=True)
print(uk_opera_df[:9]) # to compare and check that we have summed correctly
total_perform_df[:3]

        town  performances
0   Aberdeen             2
1   Aberdeen             2
2  Aldeburgh             2
3  Aldeburgh             1
4  Aldeburgh             1
5  Aldeburgh             1
6  Aldeburgh             1
7    Bampton             2
8     Bangor             1


town  performances
0   Aberdeen             4
1  Aldeburgh             6
2    Bampton             2

Knowing that our cleaned opera data matches all the towns to a lat and long, we can do a left join. Then select just the columns town, performances, latitude, longitude.

In [21]:
plot_df = total_perform_df.merge(uk_cities_df, how='left', left_on='town', right_on='name')
plot_df = plot_df[['town', 'latitude', 'longitude', 'performances']]
plot_df[:3]

town  latitude  longitude  performances
0   Aberdeen  57.14369   -2.09814             4
1  Aldeburgh  52.15492    1.60215             6
2    Bampton  51.72634   -1.54547             2

Finally lets rename the columns to match the spec in the question, and display the bottom 5 rows. (Note that the column headers are the wrong way around for latitude and longitude in the question. lat -1.25596, long 51.75222 is somewhere off the coast of Somalia...)

In [22]:
plot_df.rename(columns={'town': 'City', 'performances': 'Performances'}, inplace=True)
plot_df.tail(5)
plot_df.describe()

latitude  longitude  Performances
count  61.000000  61.000000     61.000000
mean   52.802150  -1.986571     15.327869
std     1.722479   1.749234     46.067965
min    50.265260  -7.309340      1.000000
25%    51.406060  -3.050000      2.000000
50%    52.481420  -1.899830      6.000000
75%    53.796480  -1.091250     13.000000
max    57.479080   1.602150    353.000000

#### 9. Plot the data on a map

Having created an appropriate DataFrame, you should now plot the data on a map.

As stated, your task is to plot a map of the UK, containing markers such that:
1. the location of the marker shows the location where one or more performances took place, and
2. the size of each marker represents the number of opera performances which took place at that location in the 2017/2018 opera season.

Your solution should be carried out using the folium library: you must provide Python commands which generate the required plot from your plotting DataFrame.

Finally, comment on your map: where are the main venues for opera in the UK?

*(5 marks)*



First we create a map object of a reasonable size and zoom level. A search engine and a little experimentation gives us a geographical centre of the uk (a point in Morcombe bay) and the zoom level of 6.

In [23]:
opera_map = folium.Map(location=[54.1, -2.9], width=960, height=580, zoom_start=6)

Next we can use the add_marker recipe with some modifications to plot a Circle with a relative size based on the 'Performances' column. We also add a marker that gives the name of the location('City') and the number of performances there in 2017/18.

In [24]:
def add_marker(row, fmap):
    """
    Add a circle to a map based on magnitude of performances column of dataframe.
    Increase the scale of the circle so as to have a decernable difference on a plot of the whole country.
    """
    
    lat = row['latitude']
    long = row['longitude']
    perf = row['Performances'] * 40
    city = row['City']
    
    folium.Circle(location=[lat,long], radius = perf,
                  popup=f'{perf} operas in {city} in 2017/18 season.',
                  color='red', fill_color='red',
                  fill_opacity=0.8).add_to(fmap)

Now we have a function to add a marker for each row, we apply it to the dataframe along axis 1 so as to select each row individually. Finally we can display the plot.

In [25]:
plot_df.apply(add_marker, fmap=opera_map, axis=1)
opera_map

As we can see, London had a lot more Operas than anywhere else during the 17/18 season. If this is the story we wanted to tell, this map does it well. We could scale up the size of the circles by a factor of 2 to make the point even more, but this marker size lets us see all the data, and not allow London marker to envelop Bromley.
Due to size disparity with London opera numbers, it is hard to see much of a difference between the other towns. At this zoom level everywhere else seems to have the same size circle. We can change this by scaling the size of the circles. Lets look at an overview of the data.

In [26]:
plot_df.describe()

latitude  longitude  Performances
count  61.000000  61.000000     61.000000
mean   52.802150  -1.986571     15.327869
std     1.722479   1.749234     46.067965
min    50.265260  -7.309340      1.000000
25%    51.406060  -3.050000      2.000000
50%    52.481420  -1.899830      6.000000
75%    53.796480  -1.091250     13.000000
max    57.479080   1.602150    353.000000

The maximum number of performances is 2 orders of magnitude difference from the minimum. This gives me the idea to plot the natural log of the performances as the circle size. We can use the numpy log function

In [27]:
import numpy as np
plot_df['scaledPerformances'] = plot_df['Performances'].apply(np.log, axis=1)+1 # add 1 to ensure a minimum circle size 
plot_df.head(5)

City  latitude  longitude  Performances  scaledPerformances
0   Aberdeen  57.14369   -2.09814             4            2.386294
1  Aldeburgh  52.15492    1.60215             6            2.791759
2    Bampton  51.72634   -1.54547             2            1.693147
3     Bangor  54.66079   -5.66802             1            1.000000
4     Bangor  53.22752   -4.12936             1            1.000000

Now we declare a new map, and define a new function to add this scaled size to the map. We also multiplied the scaled number so as to get larger circles to emphasise the differences with the majority of the data. We can also change the tile set to increase the contrast with our markers.

In [28]:
map_opera = folium.Map(location=[54.093409, -2.89479], width=960, height=580, zoom_start=6, tiles="CartoDB positron")
# define a new function to add each number of performances with a varying radius dependant on number of performances
def add_marker2(row, fmap):
    """Add a marker to a specific map."""
    
    lat = row['latitude']
    lon = row['longitude']
    performances = row['Performances']
    scaledPerformances = row['scaledPerformances']*2000 # increase the size of the marker to emphasize the differences
    city = row['City']
    
    folium.Circle(location=[lat, lon], radius=scaledPerformances,
                  popup=f'Operas in {city} during 2017/18 season: {performances}.', 
                  color='red', fill_color='red',
                  fill_opacity=0.8).add_to(fmap)

In [29]:
# call this function for each row in the df
plot_df.apply(add_marker2, fmap=map_opera, axis=1)
# display the new map
map_opera

As we can see this version of the map allows us to see the differences between the majority of the towns hosting operas. It does however hide the massive disparity between London and the rest of the UK. This map tells a different story. At first glance, it appears that the number of operas performed is fairly evenly spread across the UK, where as the data shows some very big differences. This goes to show that the way data is presented can have a massive impact on the interpretation and influnce,